gum/backend-windows/gummemoryaccessmonitor-windows.c

```diff
@@ -313,30 +439,86 @@ gum_memory_access_monitor_on_exception (GumExceptionDetails * details,
+  for (i = 0; i != priv->num_pages; i++)
   {
-    const GumMemoryRange * r = &priv->ranges[i];
-
-    if (GUM_MEMORY_RANGE_INCLUDES (r, GUM_ADDRESS (d.address)))
+    const GumPageDetails * page = &priv->pages_details[i];
+    const GumMemoryRange * r = &priv->ranges[page->range_index];
+    guint operation_mask;
+    guint operations_reported;
+    guint pages_remaining;
+
+    if ((page->address <= d.address) && 
+        ((guint8*) page->address + priv->page_size > (guint8*) d.address))
     {
-      d.range_index = i;
-      d.page_index = ((guint8 *) d.address -
-          (guint8 *) r->base_address) / priv->page_size;
-      d.pages_completed = priv->pages_total -
-          (g_atomic_int_add (&priv->pages_remaining, -1) - 1);
+      /* make sure that we don't misinterpret access violation / page guard */
+      if (page->is_guarded)
+      {
+        if (details->type != GUM_EXCEPTION_GUARD_PAGE)
+          return FALSE;
+      }
+      else if (details->type == GUM_EXCEPTION_ACCESS_VIOLATION)
+      {
+        GumPageProtection gum_original_protection = 
+            gum_page_protection_from_windows (page->original_protection);
+        switch (d.operation)
+        {
+        case GUM_MEMOP_READ:
+          if ((gum_original_protection & GUM_PAGE_READ) == 0)
+            return FALSE;
+          break;
+        case GUM_MEMOP_WRITE:
+          if ((gum_original_protection & GUM_PAGE_WRITE) == 0)
+            return FALSE;
+          break;
+        case GUM_MEMOP_EXECUTE:
+          if ((gum_original_protection & GUM_PAGE_EXECUTE) == 0)
+            return FALSE;
+          break;
+        default:
+          g_assert_not_reached();
+        }
+      }
+      else 
+        return FALSE;
+
+      /* restore the original protection if needed */
+      if (((priv->auto_reset) && (!page->is_guarded)))
+      {
+        DWORD old_prot;
+        /* may be called multiple times in case of simultaneous access
+         * but it should not be a problem */
+        VirtualProtect (
+            (guint8 *) d.address - (((guintptr) d.address) % priv->page_size),
+            priv->page_size, page->original_protection, &old_prot);
+      }
+
+      /* if an operation was already reported, don't report it. */
+      operation_mask = 1 << d.operation;
+      operations_reported = g_atomic_int_or (&page->completed, operation_mask);
+      if (operations_reported && priv->auto_reset)
+      {
+        return FALSE;
+      }
+
+      pages_remaining;
+      if (!operations_reported)
+        pages_remaining = g_atomic_int_add (&priv->pages_remaining, -1) - 1;
+      else
+        pages_remaining = g_atomic_int_get (&priv->pages_remaining);
+      d.pages_completed = priv->pages_total - pages_remaining;
+
+      d.range_index = page->range_index;
+      d.page_index = (guint8 *) d.address - (guint8 *) r->base_address;
+      d.page_index = d.page_index / priv->page_size;
       d.pages_total = priv->pages_total;
-
```

Comment:
Let's keep this line, it eases readability. (Separating the "prepping arguments to call that function" from "calling that function".)